File: Polynomial Regression/Polynomial_Regression/Polynomial Regression1.py
# -*- coding: utf-8 -*-
"""
Created on Thu Dec 21 12:13:41 2017

@author: Ali Baloch
"""
#Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# Just to check weather the employee is telling truth about his/her salary or bluffing
X = dataset.iloc[:, 1:2].values # transforming the column into matrix of feature so that we can apply polynomial rules
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)
y_pred = lin_reg.predict(X)

# Fitting Polynomial Regression to the Dataset
from sklearn.preprocessing import PolynomialFeatures
# poly_reg is going to be the transfroming tool that transform our matrix of feature X
#into the a new metrix of feature that we'r going to call X poly which will be a new matrix of feature
# which will contain only our independent varibale x1 but also x1^2 (x1 square)
poly_reg = PolynomialFeatures(degree = 4) # made a squre with the first column of constan 1
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)
y_pred_poly = lin_reg_2.predict(X_poly)

# Visualising the Linear Regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X),color = 'blue')
plt.title("Truth Or Bluff (Linear Regression)")
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Visualising the polynomial Regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid),1))
plt.scatter(X, y, color = 'black')
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'purple')
plt.title("Truth or Bluff (Ploynomial Regression)")
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()


# predict a new rsult with Linear Regression
lin_reg.predict(6.5)

#predict a new result with Ploynimial Regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: Polynomial Regression/Polynomial_Regression/polynomial_regression1.R
# Polynomial Regression
# Data Preprocessing 

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3] # select only level and salary column

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
#library(caTools)
#set.seed(123)
#split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
#training_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Linear Regression to the dataset
lin_reg = lm(formula = Salary ~ Level, 
             data = dataset)

#Fitting Ploynomial Regression to the dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ ., 
             data = dataset)

# Visualising the Linear Regression results
#install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color ='red') + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
             color ='blue') +
  ggtitle('Truth or Bluff (Linear Regression)') +
  xlab('Levels') +
  ylab('Salary')

# Visualising the Polynomial Regression result
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color ='red') + 
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            color ='blue') +
  ggtitle('Truth or Bluff (Linear Regression)') +
  xlab('Levels') +
  ylab('Salary')

# predict the new rsult with Linear Regression
y_pred = predict(lin_reg, data.frame(Level= 6.5))

# Preidict the new result with PPloynomial Regression
y_pred_poly = predict(poly_reg, data.frame(Level =6.5, 
                                           Level2 =6.5^2, 
                                           Level3 = 6.5^3, 
                                           Level4 = 6.5^4))
